Create rule S6365: Use symmetric transfer to switch execution between coroutines (#475)
SonarSource/rspec: rules/S6365/cfamily/rule.adoc

or to some already suspended coroutine (including the current coroutine).

The resumption of the coroutine represented by the `std::coroutine_handle` object is usually performed by calling the `.resume()` on it.
However, performing such an operation during the execution of `await_suspend` (that is part of `co_await` expression evaluation) will preserve the activation frame of the `await_suspend` function and the calling code on the stack.
This may lead to stack overflows in a situation where the chain of directly resumed coroutines is deep enough.

The use of the symmetric transfer may avoid this problem. When the `await_suspend` function returns a `std::coroutine_handle`, the compiler will automatically use this handle to resume its coroutine after `await_suspend` returns (and its activation frame is removed from the stack).
Or, when a `std::noop_coroutine_handle` is returned, the execution will be passed to the caller.

Symmetric transfer solution can also be used to resume the current coroutine (by returning handle passed as the parameter). 
However, in such cases, conditional suspension can be a more optimal solution.

> This rule raises an issue on the invocations of `std::coroutine_handle::resume` inside `await_suspend`.

Reviewer: ```suggestion
This rule raises an issue on the invocations of `std::coroutine_handle::resume` and `std::coroutine_handle::operator()` inside `await_suspend`.
```

@tomasz-kaminski-sonarsource WDYT? Should it go further and mention when it is/isn't raised w.r.t. S6366? For example, something like this:
```suggestion
This rule raises an issue on the invocations of `std::coroutine_handle::resume` and `std::coroutine_handle::operator()` inside `await_suspend` when conditional suspension (cf. S6366) isn't applicable.
```
(I'm not convinced I chose the right wording...)

Or would you say the current wording is good enough because it doesn't go too much into details?

Reviewer: As discussed with @tomasz-kaminski-sonarsource, a better option:
```suggestion
This rule raises an issue on `await_suspend` functions that could use symmetric transfer.
```

> With {cpp}20 coroutines, the `co_await`/`co_yield` expression suspends the currently executed coroutine and resumes the execution of either the caller of the coroutine function
> or to some already suspended coroutine (including the current coroutine).

Reviewer 2: ```suggestion
With {cpp}20 coroutines, the `co_await`/`co_yield` expression suspends the currently executed coroutine and resumes the execution of either the caller or the coroutine function
or to some already suspended coroutine (including the current coroutine).
```

The resumption of the coroutine represented by the `std::coroutine_handle` object is usually performed by calling the `.resume()` on it.
However, performing such an operation during the execution of `await_suspend` (that is part of `co_await` expression evaluation) will preserve the activation frame of the `await_suspend` function and the calling code on the stack.
This may lead to stack overflows in a situation where the chain of directly resumed coroutines is deep enough.

> This problem may be avoided by the use of the symmetric transfer. When the `await_suspend` function returns a `std::coroutine_handle`,
> this handle will be automatically used by the compiler to resume its coroutine after `await_suspend` returns (and its activation frame is removed from the stack).

Reviewer 2: ```suggestion
The use of the symmetric transfer may avoid this problem. When the `await_suspend` function returns a `std::coroutine_handle`, the compiler will automatically use this handle to resume its coroutine after `await_suspend` returns (and its activation frame is removed from the stack).
```

Or, when a `std::noop_coroutine_handle` is returned, the execution will be passed to the caller.

> Symmetric transfer solution can be also used to resume the current coroutine (by returning handle passed as the parameter). 

Reviewer 2: ```suggestion
Symmetric transfer solution can also be used to resume the current coroutine (by returning handle passed as the parameter). 
```

However, in such cases, conditional suspension can be a more optimal solution.

This rule raises an issue on the invocations of `std::coroutine_handle::resume` inside `await_suspend`.

== Noncompliant Code Example

----
struct InvokeOtherAwaiter {
  /* .... */
  void await_suspend(std::coroutine_handle<PromiseType> current) {
    if (auto other = current.promise().other_handle) {
      other.resume(); // Noncompliant 
    }
  }

> }

Reviewer: ```suggestion
};
```

struct WaitForAwaiter {
  Event& event;
  /* .... */
  void await_suspend(std::coroutine_handle<> current) {
    if (bool ready = event.register_callback(current)) {
      current.resume(); // Noncompliant
    }
  }

> }

Reviewer: ```suggestion
};
```

struct BufferedExecutionAwaiter {
  std::queue<std::coroutine_handle<>>& taskQueue;
  /* .... */
  void await_suspend(std::coroutine_handle<> current) {
    if (taskQueue.empty()) {
      current.resume(); // Noncompliant
    }
    auto next = taskQueue.front();
    taskQueue.pop();
    taskQueue.push(current);
    next.resume(); // Noncompliant
  }

> }

Reviewer: ```suggestion
};
```

----

== Compliant Solution

----
struct InvokeOtherAwaiter {
  /* .... */
  std::coroutine_handle<> await_suspend(std::coroutine_handle<PromiseType> current) {
    if (auto other = current.promise().other_handle) {
      return other;
    } else {
      return std::noop_coroutine();
    }
  }

> }

Reviewer: ```suggestion
};
```

struct WaitForAwaiter {
  Event& event;
  /* .... */
  std::coroutine_handle<> await_suspend(std::coroutine_handle<> current) {
    if (bool ready = event.register_callback(current)) {
      return current;
    } else {  
      return std::noop_coroutine()
    }
  }
  // Alternatively
  bool await_suspend(std::coroutine_handle<> current) {
    return !event.register_callback(current);
  }

> }

Reviewer: ```suggestion
};
```

struct BufferedExecutionAwaiter {
  std::queue<std::coroutine_handle<>>& taskQueue;
  /* .... */
  std::coroutine_handle<> await_suspend(std::coroutine_handle<> current) {
    if (taskQueue.empty()) {
       return current;
    }
    auto next = list.front();
    taskQueue.pop();
    taskQueue.push(current);
    return next;
  }

> }

Reviewer: ```suggestion
};
```

{cpp}20 coroutines, the `co_await`/`co_yield` expression suspends the currently executed coroutine and resumes the execution to either caller of coroutine function
or to some already suspended coroutine (including current coroutine).

The resumption of the coroutine represented by the `std::coroutine_handle` object, is usually performed by calling the `.resume()` on it.
However, performing such operation during execution of `await_suspend` (that is part of the `await`), will preserve the activation frame of the `await_suspend` method and the calling coode on the stack.
This may lead to stack overflow, in a situation when the chain or directly resumed coroutines is deep enough.

This problem may be avoided by the use of the symmetric transfer. If the `await_suspend` function returns a specialization `std::coroutine_handle`,
this handle will be automatically resumed by the compiler after the function returns (and its activation frame is removed from the stack).

> Or in a case when special `std::noop_coroutine_handle` is returned, the execution will be passed to the caller.

Reviewer: ```suggestion
Or, when a `std::noop_coroutine_handle` is returned, the execution will be passed to the caller.
```

> Symmetric transfer solution can be also used to resume the current coroutine (by returning handle passed as the parameter), 
> however in such cases other more optimal methods are available.

Reviewer: ```suggestion
Symmetric transfer solution can be also used to resume the current coroutine (by returning handle passed as the parameter). 
However, this is suboptimal and returning `false` will achieve the same.
```

Maybe we should motivate _why_ returning the handle parameter is not the best solution.

> This rule raises an issue on invocation of the `resume` method on `coroutine_handle` performed inside the `await_suspend` method.

Reviewer: ```suggestion
This rule raises an issue on the invocations of `std::coroutine_handle::resume` inside `await_suspend`.
```

> == Noncompliant Code Example

Reviewer: Below, I've added suggestions to rename some variables for consistency and to ensure they are defined. You may want to rename them differently -- I don't have a strong feeling about their names.

----
struct InvokeOtherAwaiter {

>   std::coroutine_handle<> other;

Reviewer: ```suggestion
  std::coroutine_handle<> other_handle;
```

  /* .... */
  void await_suspend(std::coroutine_handle<PromiseType> current) {

>     if (auto other = h.promise().other_handle) {

Reviewer: ```suggestion
    if (auto other = current.promise().other_handle) {
```

>       other.resume(); // Noncomplaint 

Reviewer: ```suggestion
      other.resume(); // Noncompliant 
```

    }
  }
}

struct WaitForAwaiter {
  Event& event;
  /* .... */
  void await_suspend(std::coroutine_handle<> current) {
    if (bool ready = event.register_callback(current)) {

>       h.resume(); // Noncomplaint

Reviewer: ```suggestion
      h.resume(); // Noncompliant
```
BTW: `h` isn't defined.

>       h.resume(); // Noncompliant

Reviewer: ```suggestion
      current.resume(); // Noncompliant
```

    }
  }
}

struct BufferedExecutionAwaiter {
  std::queue<std::coroutine_handle<>>& taskQueue;
  /* .... */

>   void await_suspend(std::coroutine_handle<> h) {

Reviewer: ```suggestion
  void await_suspend(std::coroutine_handle<> current) {
```
for consistency

    if (taskQueue.empty()) {

>       h.resume(); // Noncomplaint

Reviewer: ```suggestion
      h.resume(); // Noncompliant
```

>       h.resume(); // Noncompliant

Reviewer: ```suggestion
      current.resume(); // Noncompliant
```

    }

>     auto next = list.front();

Reviewer: ```suggestion
    auto next = taskQueue.front();
```
I suppose.

    taskQueue.pop();

>     taskQueue.push(h);

Reviewer: ```suggestion
    taskQueue.push(current);
```

    next.resume();
  }
}
----

== Compliant Solution

----

> struct InvokeOtherAwaiter {

Reviewer: The examples are not in the same order in the two sections.

Reviewer 3: Reorderd

>   std::coroutine_handle<> other;

Reviewer: ```suggestion
  std::coroutine_handle<> other_handle;
```

  /* .... */

>   std::coroutine_handle<> await_suspend(std::coroutine_handle<PromiseType> h) {

Reviewer: ```suggestion
  std::coroutine_handle<> await_suspend(std::coroutine_handle<PromiseType> current) {
```

>     if (auto other = h.promise().other_handle) {

Reviewer: ```suggestion
    if (auto other = current.promise().other_handle) {
```

      return other;
    } else {
      return std::noop_coroutine();
    }
  }
}

struct WaitForAwaiter {
  Event& event;
  /* .... */

>   std::coroutine_handle<> await_suspend(std::coroutine_handle<> v) {

Reviewer: ```suggestion
  std::coroutine_handle<> await_suspend(std::coroutine_handle<> current) {
```

    if (taskQueue.empty()) {
       return h;
    }
    auto next = list.front();
    taskQueue.pop();
    taskQueue.push(h);
    return next;
  }
}

struct WaitForAwaiter {
  Event& event;
  /* .... */

>   void await_suspend(std::coroutine_handle<> v) {

Reviewer: ```suggestion
  auto await_suspend(std::coroutine_handle<> v) {
```
or, if `auto` can't work here,
```suggestion
  std::coroutine_handle<> await_suspend(std::coroutine_handle<> v) {
```

Reviewer 3: I think showing return type here add to the example, and makes it cleaner.

    if (bool ready = event.register_callback(current)) {
      return current;
    } else {  

>       return std:noop_coroutine()

Reviewer: ```suggestion
      return std::noop_coroutine()
```

>        return h;

Reviewer: ```suggestion
       return current;
```

    }
    auto next = list.front();
    taskQueue.pop();

>     taskQueue.push(h);

Reviewer: ```suggestion
    taskQueue.push(current);
```

    return !event.register_callback(current);
  }
}
----

== See

> S6366 - conditionally suspending current coroutine in optimal way  

Reviewer: Is this number already reserved or do we need to verify it won't be used by another rule by the time this PR gets merged?

Reviewer 3: I already reserved it, this is why it was fixme before.

With {cpp}20 coroutines, the `co_await`/`co_yield` expression suspends the currently executed coroutine and resumes the execution of either the caller of the coroutine function
or to some already suspended coroutine (including the current coroutine).

The resumption of the coroutine represented by the `std::coroutine_handle` object is usually performed by calling the `.resume()` on it.

> However, performing such an operation during the execution of `await_suspend` (that is part of `co_await` expression evaluation), will preserve the activation frame of the `await_suspend` function and the calling code on the stack.

Reviewer: ```suggestion
However, performing such an operation during the execution of `await_suspend` (that is part of `co_await` expression evaluation) will preserve the activation frame of the `await_suspend` function and the calling code on the stack.
```

This may lead to stack overflows in a situation where the chain of directly resumed coroutines is deep enough.

This problem may be avoided by the use of the symmetric transfer. When the `await_suspend` function returns a `std::coroutine_handle`,
this handle will be automatically used by the compiler to resume its coroutine after `await_suspend` returns (and its activation frame is removed from the stack).
Or, when a `std::noop_coroutine_handle` is returned, the execution will be passed to the caller.

Symmetric transfer solution can be also used to resume the current coroutine (by returning handle passed as the parameter). 

> However, in such case, conditional suspension can more optimal solution.

Reviewer: ```suggestion
However, in such cases, conditional suspension can be a more optimal solution.
```
